adserver/validators.py

```diff
@@ -115,3 +116,57 @@ def _validate_strs(self, words):
         for word in words:
             if not isinstance(word, str):
                 raise ValidationError(self.messages["str"], params={"word": word})
+
+
+@deconstructible
+class TrafficFillValidator(TargetingParametersValidator):
+
+    """
+    Validator for traffic fill and cap data.
+
+    eg.
+      {
+        "publishers": {"publisher1": 0.1, "publisher2": 0.05},
+        "countries": {"us": 0.1, "ca": 0.05, "de": 0.05},
+        "regions": {"us-ca": 0.25, "eu": 0.5},
+      }
+    """
+
+    code = "traffic-validator"
+
+    validators = {
+        "countries": "_validate_countries",
+        "regions": "_validate_regions",
+        "publishers": "_validate_publishers",
+    }
+
+    def _validate_countries(self, country_values):
+        for code, value in country_values.items():
+            if code not in self.country_set:
+                raise ValidationError(
+                    self.messages["country_code"], params={"code": code}
+                )
+
+            if not isinstance(value, float) or value > 1.0 or value < 0:
+                raise ValidationError(self.messages["percent"], params={"value": value})
+
+    def _validate_regions(self, region_values):
+        from .models import Region  # noqa
+
+        regions = Region.load_from_cache()
+        for slug, value in region_values.items():
+            if slug not in regions:
+                raise ValidationError(self.messages["region"], params={"value": slug})
+
+            if not isinstance(value, float) or value > 1.0 or value < 0:
+                raise ValidationError(self.messages["percent"], params={"value": value})
+
+    def _validate_publishers(self, publisher_values):
+        for publisher_slug, value in publisher_values.items():
+            if not isinstance(publisher_slug, str):
```

Comment:
Do we actually want to validate it's a valid publisher slug? Not super important, but I did accidentally set one wrong the other day, but caught it once it was displayed back to me in the flight admin.

Reply:
We don't in the targeting validator. It would cause an extra query to do that validation.

Reply:
I think the query is fine, since it's only validated on save, but I'm fine not validating it.